clang/lib/Sema/SemaDecl.cpp

```diff
@@ -6970,6 +6970,18 @@ static void checkAttributesAfterMerging(Sema &S, NamedDecl &ND) {
         }
       }
     }
+
+    for (unsigned int I = 0; I < FD->getNumParams(); ++I) {
+      const ParmVarDecl *P = FD->getParamDecl(I);
+
+      // The [[lifetimebound]] attribute can be applied to a function parameter
+      // only if the function returns a value.
+      if (auto *A = P->getAttr<LifetimeBoundAttr>()) {
```

Comment:
An interesting case that [came up](https://github.com/llvm/llvm-project/issues/107556#issuecomment-2333906528) involves dependent return type.

I think we should not be showing this error if there is a dependent return type.
```cpp
template <class T, class R = void> R foo(const T& t [[clang::lifetimebound]]);
foo<int>(1);
const int& r = foo<int, int>(1);
```

There is no way to fix this for one instantiation and not for the other so we should be accepting these.

Can you also add test involving tempates with and without dependent return types.

Reply:
I believe we don't.
I've checked and the return type is TemplateTypeParm so it is not considered to be void.

Reply:
I think if the type is dependent, the check should be skipped.
Hopefully the check is then run upon instantiation

Reply:
Added a test to demonstrate that.